examples/beamer_location_builder/lib/main.dart

```diff
@@ -0,0 +1,170 @@
+            .map((book) => ListTile(
+                  title: Text(book['title']),
+                  subtitle: Text(book['author']),
+                  onTap: () => Beamer.of(context).currentLocation.update(
+                        (state) => state.copyWith(
+                          pathBlueprintSegments: ['books', ':bookId'],
+                          pathParameters: {'bookId': book['id']},
+                        ),
+                      ),
+                ))
+            .toList(),
+      ),
+    );
+  }
+}
+
+class BookDetailsScreen extends StatelessWidget {
+  BookDetailsScreen({this.bookId});
+
+  final String bookId;
+
+  @override
+  Widget build(BuildContext context) {
+    final books = context.read<Books>().books;
+    final book = books.firstWhere((book) => book['id'] == bookId);
+    return Scaffold(
+      appBar: AppBar(
+        title: Text(book['title'] +
+            (' (I also' +
+                (books != null ? '' : " don't") +
+                ' have access to books)')),
+      ),
+      body: Padding(
+        padding: const EdgeInsets.all(8.0),
+        child: Text('Author: ${book['author']}'),
+      ),
+    );
+  }
+}
+
+// LOCATIONS
+class HomeLocation extends BeamLocation {
+  HomeLocation(BeamState state) : super(state);
+
+  @override
+  List<String> get pathBlueprints => ['/'];
+
+  @override
+  List<BeamPage> pagesBuilder(BuildContext context, BeamState state) => [
+        BeamPage(
+          key: ValueKey('home'),
+          child: HomeScreen(),
+        ),
+      ];
+}
+
+class BooksLocation extends BeamLocation {
+  BooksLocation(BeamState state) : super(state);
+
+  @override
+  Widget builder(BuildContext context, Widget navigator) =>
+      ChangeNotifierProvider(
+        create: (context) => Books(),
+        child: navigator,
+      );
+
+  @override
+  List<String> get pathBlueprints => ['/books/:bookId'];
+
+  @override
+  List<BeamPage> pagesBuilder(BuildContext context, BeamState state) {
+    return [
+      ...HomeLocation(state).pagesBuilder(context, state),
+      if (state.uri.pathSegments.contains('books'))
+        BeamPage(
+          key: ValueKey('books-${state.queryParameters['title'] ?? ''}'),
+          child: BooksScreen(),
+        ),
+      if (state.pathParameters.containsKey('bookId'))
+        BeamPage(
+          key: ValueKey('book-${state.pathParameters['bookId']}'),
+          child: BookDetailsScreen(
+            bookId: state.pathParameters['bookId'],
+          ),
+        ),
+    ];
+  }
+}
+
+// APP
+class MyApp extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp.router(
+      debugShowCheckedModeBanner: false,
+      routerDelegate: BeamerRouterDelegate(
+        locationBuilder: BeamerLocationBuilder(
+          beamLocations: (state) => [
+            BooksLocation(state),
+            HomeLocation(state),
+          ],
+        ),
```

Comment:
The initial goal of making this an array is not possible right now, because the `state` constructor property is required for each `BeamLocation`. 
This could be solved by making this an optional property like this:

```dart
abstract class BeamLocation<T extends BeamState> extends ChangeNotifier {
  BeamLocation([T? state]) {
    _state = createState(state ?? BeamState());
  }

...
}
```

Then we can e.g. simply call `BooksLocation()` with no state.
@slovnicki What do you think?

Reply:
I was aiming to ease the boilerplate, because without specifying zero-argument constructor in `BeamLocation `, IDE will provide a "quick fix" and make it for every class that extends it.

But, on the other hand, it really makes no sense creating all the `BeamLocation`s with same state when using your builders. Furthermore, anyone **not** using your builders is probably an experienced enough user to know why separate constructors need to be created.

So I agree and will make it optional.